flow/scene_update_context.cc

```diff
@@ -284,43 +281,38 @@ SceneUpdateContext::Transform::~Transform() {
   context().top_scale_y_ = previous_scale_y_;
 }
 
-SceneUpdateContext::Shape::Shape(SceneUpdateContext& context)
-    : Entity(context), shape_node_(context.session()) {
-  entity_node().AddChild(shape_node_);
+SceneUpdateContext::Clip::Clip(SceneUpdateContext& context,
+                               const SkRect& shape_bounds)
+    : Entity(context) {
+  SetEntityNodeClipPlanes(entity_node(), shape_bounds);
 }
 
 SceneUpdateContext::Frame::Frame(SceneUpdateContext& context,
                                  const SkRRect& rrect,
                                  SkColor color,
-                                 float local_elevation,
-                                 float world_elevation,
-                                 float depth,
+                                 float opacity,
+                                 float elevation,
                                  Layer* layer)
-    : Shape(context),
+    : Entity(context),
+      opacity_node_(context.session()),
+      shape_node_(context.session()),
+      layer_(layer),
       rrect_(rrect),
-      color_(color),
       paint_bounds_(SkRect::MakeEmpty()),
-      layer_(layer) {
-  if (depth > -1 && world_elevation > depth) {
-    // TODO(mklim): Deal with bounds overflow more elegantly. We'd like to be
-    // able to have developers specify the behavior here to alternatives besides
-    // clamping, like normalization on some arbitrary curve.
-
-    // Clamp the local z coordinate at our max bound. Take into account the
-    // parent z position here to fix clamping in cases where the child is
-    // overflowing because of its parents.
-    const float parent_elevation = world_elevation - local_elevation;
```

Comment:
Yeah, correct that we still need it. At a glance it looks like this functionality has been moved into `ClampElevation` in this PR. From what I can tell it seems like that should be called and clamping `-elevation` before it gets to this call, but it's hard to tell for sure and I'd be worried that down the line someone would modify `elevation` here without realizing that it shouldn't be moved post-clamp. I'll add this to my queue to review more thoroughly.

Reply:
Yep, I moved this logic into ClampElevation, because we are trying to reduce Fuchsia-specific code paths.  A future PR is going to make these objects like Frame stateless, so we can make elevation const here if you want.

Reply:
Yep!  I moved this into ClampElevation because we are trying to reduce the number of Fuchsia-specific code paths.  My next PR is going to make these objects like SceneUpdateContext::Frame stateless so we can make elevation const here if you want

Reply:
Just got a chance to go through this in more detail. I have one question but other than that the elevation changes LG.

I don't think `const` would be super useful here since it's generally used as a language feature to make the variable itself immutable. It's not the `elevation` param specifically that matters here but more the larger context of where this frame is elevated shouldn't be altered at this point. I'm not sure the right way to convey that. It may be helpful to change the param name in this function from `elevation` to `z_position` and have `UpdateScene` pass in the `-elevation` itself, so that this function isn't expected to use a modified version of this value at all. Or leave a comment. But I don't think it's a huge risk either way.